### DISCUSSION TITLE
Can no longer run Fred's ImageMagick Scripts - curves; get sed: RE error: illegal byte sequence

### DISCUSSION THREAD

**Discussion by PacoH** [ImageMagick/ImageMagick#6829] [Help]
### ImageMagick version

7.1.1-21

### Operating system, version and so on

OS X Ventura

### Description

This just happened today. I can no longer run `curves` without getting `sed: RE error: illegal byte sequence`.

**Comment by fmw42** [comment]
It works fine for me on IM 7.1.1.21 Mac OSX Ventura.  Please provide your input image and exact command line.

**Reply by PacoH** [reply]
Thanks for your speedy reply!

```
curves -h


curves:
sed: RE error: illegal byte sequence
```
and any other command using `curves`.

I just found this at [regex - RE error: illegal byte sequence on Mac OS X - Stack Overflow](https://stackoverflow.com/questions/19242275/re-error-illegal-byte-sequence-on-mac-os-x/19770395#19770395):
```
export LC_CTYPE=C 
export LANG=C
```
This worked but I have never had to do this before.

Reading further, I see:
>This will completely disable support for Unicode characters in your shell. Goodbye emojis, fancy line drawing characters, letters with accents, .... Much better to just set this for the sed command only, as described in other answers.

I added the export commands at the top of `curves` and that works. I don't know if this will cause problems later but I think it is confined to the current terminal window.

**Reply by fmw42** [reply]
For some reason that I do not understand, bash in OSX Ventura did not like the copyright symbol that I used at the top of the script.  Try this and let me if it works for you.  It seems to work for me.

[curves.zip](https://github.com/ImageMagick/ImageMagick/files/13199854/curves.zip)

**Reply by PacoH** [reply]
Thanks.

That worked but I have been using the original version for weeks now without incident and I have not upgraded Ventura yet for fear it will break yet more things as it always does. I also rebooted the system and cleared the PRAM.

What could be the cause of this? I did use `brew` to install and upgrade things. Could that have done something?

**Reply by fmw42** [reply]
I have no idea.  It is very strange. I do not think brew has anything to do with it. I have never used Homebrew.

**Reply by PacoH** [reply]
I found this:

**[Bug in sed | Apple Developer Forums](https://developer.apple.com/forums/thread/720787)**

>With a UTF-8 locale, this sed command to insert a character at the beginning of each line incorrectly crashes with an error about an illegal byte sequence:
```
$ echo “hi | LANG=en_US.UTF-8 sed -e s'/^/x/g'
sed: RE error: illegal byte sequence
```
>If you change the ^ to an ordinary character, or drop the g flag, it works fine.

I verified it. 

The problem goes away with GNU sed:

```
echo “hi | LANG=en_US.UTF-8 gsed -e s'/^/x/g'
x“hi
```

What puzzles me is why Apple has never adopted GNU sed as the default. You have to install it yourself.

BTW, `brew` is the best way to install packages on OS X. It automatically installs dependencies and  links packages to `/usr/local/bin/` and warns you if you need to force link a package that failed to link. 

To install `gsed`:

`brew install gnu-sed`

I found a simple way to make `gsed` the default sed.

**[macos - How to use GNU sed on Mac OS 10.10+, 'brew install --default-names' no longer supported - Stack Overflow](https://stackoverflow.com/questions/30003570/how-to-use-gnu-sed-on-mac-os-10-10-brew-install-default-names-no-longer-su/31080313#31080313)**

If you need to use gnu-sed command with their normal names, you can add a "gnubin" directory to your PATH from your bashrc. Just use the following command in your bash or terminal.

`export PATH="/usr/local/opt/gnu-sed/libexec/gnubin:$PATH"`

I added this to `~/.bash_profile` and when I sourced it, the original `curves` with the `©` symbol just ran.

**Comment by fmw42** [comment]
I do not understand why the © symbol fails in my curves script. But it works fine in this script

```
#!/bin/bash

case "$1" in
-h|-help)    # help information
echo ""
echo "This is a test of a © in a file" | sed -e 's/a/the/g'
esac

exit 0

```

**Reply by PacoH** [reply]
I tried this and it does work but so does your original `curves` script since I switched to `gsed` permanently.

**Comment by fmw42** [comment]
This also works.  I just copied the top of my curves script.  So I have not found what interaction is causing this.

```
#!/bin/bash

# Copyright © Fred Weinhaus


PROGNAME=`type $0 | awk '{print $3}'`  # search for executable on path
PROGDIR=`dirname $PROGNAME`            # extract directory of program
PROGNAME=`basename $PROGNAME`          # base name of program
usage1() 
	{
	echo >&2 ""
	echo >&2 "$PROGNAME:" "$@"
	sed >&2 -e '1,/^####/d;  /^###/g;  /^#/!q;  s/^#//;  s/^ //;  4,$p' "$PROGDIR/$PROGNAME"
	}
usage2() 
	{
	echo >&2 ""
	echo >&2 "$PROGNAME:" "$@"
	sed >&2 -e '1,/^####/d;  /^######/g;  /^#/!q;  s/^#*//;  s/^ //;  4,$p' "$PROGDIR/$PROGNAME"
	}

# function to report error messages, usage and exit
errMsg()
	{
	echo ""
	echo $1
	echo ""
	usage1
	exit 1
	}


if [ $# -eq 0 ]
	then
	# help information
   echo ""
   usage2
   exit 0
elif [ $# -gt 1 ]
	then
	errMsg "--- TOO MANY ARGUMENTS WERE PROVIDED ---"
else
	while [ $# -gt 0 ]
		do
			# get parameter values
			case "$1" in
		  -h|-help)    # help information
					   echo ""
					   usage2
					   exit 0
					   ;;
			esac
			shift   # next option
	done
fi

exit 0

```

**Reply by fmw42** [reply]
Other scripts such as `cosinebell -h` work fine with the © symbol.  So I do not understand why curves has trouble.  

If I comment out these 3 lines in curves (containing ©), then it works also.

```
PROGNAME=`type $0 | awk '{print $3}'`  # search for executable on path
PROGDIR=`dirname $PROGNAME`            # extract directory of program
PROGNAME=`basename $PROGNAME`          # base name of program

```

**Reply by PacoH** [reply]
Maybe it's the combination of the `©` symbol and the statement 
```
im_version=`convert -list configure | \
	sed '/^LIB_VERSION_NUMBER */!d;  s//,/;  s/,/,0/g;  s/,0*\([0-9][0-9]\)/\1/g' | head -n 1`
```

**Reply by fmw42** [reply]
Yes an any number of changes seem to fix it including commenting that line out and replacing it with a fixed value for the IM version.  But I have that version test in other scripts and it does not fail with that error message.

**Reply by PacoH** [reply]
I tell you, OS X Ventura is the worst version of OS X to come along. It has broken so many things that it's almost criminal. Millions of user/victims have been outraged by this. I think the developers are a bunch of green newbies. It has seriously slowed down my workflow and caused me to spend, literally, days searching for solutions, or workarounds, for all the problems it has caused.

**Comment by PacoH** [comment]
The only thing I can think of is the statement
```
im_version=`convert -list configure | \
	sed '/^LIB_VERSION_NUMBER */!d;  s//,/;  s/,/,0/g;  s/,0*\([0-9][0-9]\)/\1/g' | head -n 1`
```
According to **[Bug in sed | Apple Developer Forums](https://developer.apple.com/forums/thread/720787)**:

>With a UTF-8 locale, this sed command to insert a character at the beginning of each line incorrectly crashes with an error about an illegal byte sequence:

```
echo “hi | LANG=en_US.UTF-8 sed -e s'/^/x/g'
sed: RE error: illegal byte sequence
```
>If you change the ^ to an ordinary character, or drop the g flag, it works fine. I’m guessing the g makes it check the line again, but it gets messed up trying to find the start of the line with the multi-byte character there.

I see both characters in the above-quoted statement  from `curves`.

**Comment by PacoH** [comment]
I think I found what is causing the error:
```
$ bash -x curves -h
+ xscale=100
+ yscale=100
+ channel=
+ height=10
+ precision=1
+ pinc=1
+ graphmode=
+ crad=2
+ graph=view
+ dir=.
++ type curves
++ awk '{print $3}'
+ PROGNAME=/usr/local/bin/curves
++ dirname /usr/local/bin/curves
+ PROGDIR=/usr/local/bin
++ basename /usr/local/bin/curves
+ PROGNAME=curves
+ '[' 1 -eq 0 ']'
+ '[' 1 -gt 9 ']'
+ '[' 1 -gt 0 ']'
+ case "$1" in
+ echo ''

+ usage2
+ echo ''

+ echo curves:
curves:
+ sed -e '1,/^####/d;  /^######/g;  /^#/!q;  s/^#*//;  s/^ //;  4,$p' /usr/local/bin/curves
sed: RE error: illegal byte sequence
```

Line 152:
`sed >&2 -e '1,/^####/d;  /^######/g;  /^#/!q;  s/^#*//;  s/^ //;  4,$p' "$PROGDIR/$PROGNAME"`

If I comment it out, there is no error.

**Reply by fmw42** [reply]
Yes, I found that also. But many other edits also make it work.  So I am still puzzled. See my script above. I took the top of my curves script and dropped most of it. It has that line in it, but it works.

**Reply by PacoH** [reply]
It does work but maybe it gets flagged due to what follows in the original `curves`. I sometimes get errors from syntax errors of commands that are far below otherwise working code.

**Reply by fmw42** [reply]
True, but I do not know what is the link between the issue and what is below.  If you remove the © symbol at the top it works also. There are many things that can be removed that makes it work.  I just do not see anything below that is not used in other scripts that are working or anything that catches my eye as being non-standard for my scripts.

**Reply by PacoH** [reply]
That is bizarre. What puzzles me is how this suddenly occurred after long usage. I am still using OS X Ventura 13.3.1.

**Reply by fmw42** [reply]
I cannot confirm that it has not been there for a long time.  Sorry, I have not tested that for a while. It does not happen except for the -h option.  Is the word "help" a special word?

**Reply by PacoH** [reply]
It seems to be due to a combination of different things. Which Ventura are you using?

**Comment by fmw42** [comment]
13.6
